📖 Proposal: Improving status in CAPI resources (#10897)
kubernetes-sigs/cluster-api: docs/proposals/improve-status-in-CAPI-resources.md

| `OwnerRemediated`      |                                                                                                                                                                                                                                                                |
| `Deleted`              | True if Machine is deleted; Reason can be used to observe the cleanup progress when the resource is deleted                                                                                                                                                    |
| `Paused`               | True if the Machine or the Cluster it belongs to are paused                                                                                                                                                                                                    |

> To better evaluate proposed changes, below you can find the list of current Machine's conditions:

> > Ready, InfrastructureReady, BootstrapReady, NodeHealthy, PreDrainDeleteHookSucceeded, VolumeDetachSucceeded, DrainingSucceeded.

Author: Resolving, https://github.com/kubernetes-sigs/cluster-api/issues/11126 is going to address this for both old and new conditions

| `Conditions`                      | `BackCompatibilty.Conditions` (renamed) (deprecated)     | (removed)                                          |
| `ExperimentalConditions` (new)    | `Conditions` (renamed)                                   | `Conditions`                                       |
| other fields...                   | other fields...                                          | other fields...                                    |

TODO: double check usages of status.ready.

#### KubeadmControlPlane (New)Conditions

| Condition               | Note                                                                                                                                                                                                                                                                      |
|-------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | `Available`             | True if the control plane can be reached, `EtcdClusterAvailable` is true, and `CertificatesAvailable` is true                                                                                                                                                             |

Reviewer: does this mean that 'Available' can be true even though 'MachinesAvailable' is false?

why not include 'MachinesAvailable' here?

Reviewer 2: I think because not all Machines have to be available for the control plane to be Available (we have redundancy there)

  the `ScalingDown` condition.
- As of today MachinePool does not have a notion similar to MachineDeployment's MaxUnavailability.

#### MachinePool Print columns

| Current           | To be                  |
|-------------------|------------------------|
| `NAME`            | `NAME`                 |
| `CLUSTER`         | `CLUSTER`              |
| `DESIRED` (*)     | `PAUSED` (new) (*)     |
| `REPLICAS`        | `DESIRED`              |
| `PHASE` (deleted) | `CURRENT` (*)          |
| `AGE`             | `READY`                |
| `VERSION`         | `AVAILABLE` (new)      |
|                   | `UP-TO-DATE` (renamed) |
|                   | `AGE`                  |
|                   | `VERSION`              |

(*) visible only when using `kubectl get -o wide`

Notes:
- Print columns are not subject to any deprecation rule, so it is possible to iteratively improve print columns without waiting for the next API version.
- During the implementation we are going to verify the resulting layout and eventually make final adjustments to the column list.

> ### Changes to Cluster API contract

Reviewer 3: I haven't read the details here in depth, but thank you very much for highlighting these changes in a dedicated section!

Author: Resolving, no additional feedback so far so I assume it is ok

- Disambiguate the usage of the ready term by renaming fields used for the initial provisioning workflow
- Remove `FailureReason` and `FailureMessage` to get rid of the confusing concept of terminal failures
- Transition to new, improved, K8s API conventions aligned conditions

Below you can find the relevant fields in MachineStatus v1beta2, after v1beta1 removal (end state);
Below the Go types, you can find a summary table that also shows how changes will be rolled out according to K8s deprecation rules.

```golang
type MachineStatus struct {

    // Initialization provides observations of the Machine initialization process.
    // NOTE: Fields in this struct are part of the Cluster API contract and are used to orchestrate initial Machine provisioning.
    // The value of those fields is never updated after provisioning is completed.
    // Use conditions to monitor the operational state of the Machine.
    // +optional
    Initialization *MachineInitializationStatus `json:"initialization,omitempty"`
    
    // Conditions represent the observations of a Machine's current state.
    // +optional
    // +listType=map
    // +listMapKey=type

>     Conditions []metav1.Condition `json:"conditions,omitempty"`

Author: +1
noted in the [tracking issue](https://github.com/kubernetes-sigs/cluster-api/issues/11105) for the implementation

  condition at Machine level as well as adding a new `MinReadySeconds` field (see [changes to Machine.Spec](#machine-spec))
  that will be used to compute this condition.
- Similarly, this proposal is standardizing the concept of Machine's `UpToDate` condition, however in this case it will be up to
  the Machine's owner controllers to set this condition.
- Conditions like `NodeReady` and `NodeHealthy` which depend on the connection to the remote cluster will benefit
  from the new `RemoteConnectionProbe` condition at cluster level (see [Cluster (New)Conditions](#cluster-newconditions));
  more specifically those condition should be set to `Unknown` when after `lastRemoteConnectionProbeTime` plus the value
  defined in the `--remote-conditions-grace-period` flag.
- `HealthCheckSucceeded` and `OwnerRemediated` (or `ExternalRemediationRequestAvailable`) conditions are set by the
  MachineHealthCheck controller in case a MachineHealthCheck targets the machine.
- KubeadmControlPlane also adds additional conditions to Machines, but those conditions are not included in the table above

>   for sake of simplicity (however they are documented in the KubeadmControlPlane paragraph).

Author: This is MHC behaviour not impacted by this proposal, but I think it is not an issue.
- If MHC is removed after marking a machine as unhealthy, the machine gets remediated
- If MHC is removed after marking a machine healthy, the machine it considered healthy until another MHC instance take over

type MachineReadinessGate struct {
    // ConditionType refers to a condition in the Machine's condition list with matching type.
    // Note: Both Cluster API conditions or conditions added by 3rd party controllers can be used as readiness gates.
    ConditionType string `json:"conditionType"`
}
```

| v1beta1 (tentative Dec 2024) | v1Beta2 (tentative Apr 2025) | v1beta2 after v1beta1 removal (tentative Apr 2026) |
|------------------------------|------------------------------|----------------------------------------------------|
| `MinReadySeconds` (new)      | `MinReadySeconds`            | `MinReadySeconds`                                  |
| `ReadinessGates` (new)       | `ReadinessGates`             | `ReadinessGates`                                   |
| other fields...              | other fields...              | other fields...                                    |

Notes:
- Both `MinReadySeconds` and `ReadinessGates` should be treated as other in-place propagated fields (changing them should not trigger rollouts).
- Similarly to Pod's `ReadinessGates`, also Machine's `ReadinessGates` accept only conditions with positive polarity;

>   The Cluster API project might revisit this in the future to stay aligned with Kubernetes or if there are use cases justifying this change.

Author: Adding support for negative polarity conditions is tracked in https://github.com/kubernetes-sigs/cluster-api/issues/11105 as a follow up - after we got a first implementation up and running

- Cluster API is going to maintain a `lastRemoteConnectionProbeTime` and use it in combination with the
  `--remote-connection-grace-period` flag to avoid flakes on `RemoteConnectionProbe`.
- Similarly to `lastHeartbeatTime` in Kubernetes conditions, also `lastRemoteConnectionProbeTime` will not surface on the
  API in order to avoid costly, continuous reconcile events.

#### Cluster Spec

Cluster's spec is going to be improved to allow 3rd parties to extend the semantic of the new Cluster's `Available` condition.

Below you can find the relevant fields in ClusterSpec v1beta2, after v1beta1 removal (end state);
Below the Go types, you can find a summary table that also shows how changes will be rolled out according to K8s deprecation rules.

```golang
type ClusterSpec struct {
    // AvailabilityGates specifies additional conditions to include when evaluating Cluster availability.
    // A Cluster is available if:
    // * Cluster's `RemoteConnectionProbe` and `TopologyReconciled` conditions are true and
    // * the control plane `Available` condition is true and
    // * all worker resource's `Available` conditions are true and

>     // * all conditions defined in AvailabilityGates are true as well.

Author: tracked in https://github.com/kubernetes-sigs/cluster-api/issues/11105, so we can make a pass on all the fields (might be relevant also for https://github.com/kubernetes-sigs/cluster-api/issues/10852)

      - [MachineSet (New)Conditions](#machineset-newconditions)
      - [MachineSet Print columns](#machineset-print-columns)
    - [Changes to MachineDeployment resource](#changes-to-machinedeployment-resource)
      - [MachineDeployment Status](#machinedeployment-status)
      - [MachineDeployment (New)Conditions](#machinedeployment-newconditions)
      - [MachineDeployment Print columns](#machinedeployment-print-columns)
    - [Changes to Cluster resource](#changes-to-cluster-resource)
      - [Cluster Status](#cluster-status)
        - [Cluster (New)Conditions](#cluster-newconditions)
      - [Cluster Spec](#cluster-spec)
      - [Cluster Print columns](#cluster-print-columns)
    - [Changes to KubeadmControlPlane (KCP) resource](#changes-to-kubeadmcontrolplane-kcp-resource)
      - [KubeadmControlPlane Status](#kubeadmcontrolplane-status)
      - [KubeadmControlPlane (New)Conditions](#kubeadmcontrolplane-newconditions)
      - [KubeadmControlPlane Print columns](#kubeadmcontrolplane-print-columns)
    - [Changes to MachinePool resource](#changes-to-machinepool-resource)
    - [Changes to Cluster API contract](#changes-to-cluster-api-contract)

# Summary

> This documents defines how status in CAPI resources is going to evolve in the v1beta2 API version, with the goal of

Reviewer 4: ```suggestion
This documents defines how status across all Cluster API resources is going to evolve in the v1beta2 API version, focusin on
```

improving usability and consistency across different resources in CAPI and with the rest of the ecosystem.

# Motivation

> The Cluster API community recognize that nowadays Cluster API and Kubernetes users are rightfully focused on

Reviewer 4: ```suggestion
The Cluster API community recognizes that nowadays most users are rightfully focused on
```

> building higher systems and great applications on top on those platforms, which is great.

Reviewer 4: ```suggestion
building higher level systems, offerings, and applications on these platforms.
```

> However, as a consequence of this shifted focus, most of the users don’t have time to become deep expert of Cluster API

Reviewer 4: ```suggestion
However, as the focus shifted away, most of the users don’t have time to become deep experts on Cluster API.
```

> like the first wave of adopters, and also Cluster API maintainers would like they don’t have to.

Reviewer 4: ```suggestion
```

> The effect of the trend above is the blurring of the lines not only between different Cluster API components, but also

Reviewer 4: ```suggestion
This trend is blurring the lines between different Cluster API components; between Cluster API and Kubernetes, and tools like Helm, Flux, Argo, and so on.
```

> between Cluster API, core Kubernetes and a few other broadly adopted tools like Helm or Flux (and to some extents,
> also with many others awesome tools in the ecosystem).

Reviewer 4: ```suggestion
```

> This is why Cluster API status must become simpler to understand for users, and also more consistent not only across
> different CAPI resources, but with Kubernetes core and ideally with the entire ecosystem.

Reviewer 4: ```suggestion
This proposal focused on Cluster API's resource status which must become simpler to understand, more consistent with Kubernetes, and ideally with the entire ecosystem.
```

### Goals

- Review and standardize the usage of the concept of readiness across Cluster API resources.
    - Drop or amend improper usage of readiness
    - Make the concept of Machine readiness extensible, thus allowing providers or external system to inject their readiness checks.
- Review and standardize the usage of the concept of availability across Cluster API resources.
    - Make the concept of Cluster Availability extensible, thus allowing providers or external system to inject their availability checks.
- Bubble up more information about both CP and worker Machines, ensuring consistent way across Cluster API resources.
    - Standardize replica counters and bubble them up to the Cluster resource.
    - Standardize control plane, MachineDeployment, Machine pool availability, and bubble them up to the Cluster resource.
- Introduce missing signals about connectivity to workload clusters, thus enabling to mark all the conditions
  depending on such connectivity being working with status Unknown after a certain amount of time.
- Introduce a cleaner signal about Cluster API resources lifecycle transitions, e.g. scaling up or updating.

> - Ensure everything in status can be used as a signal informing monitoring tools/automation on top of Cluster 

Reviewer 5: This is great goal. As a service provider I want to be able to use conditions as SLIs to measure that my SLOs are met.

Reviewer 2: Agree. One easy path to that should be: conditions => state metrics => some simple dashboards / alerting rules

Author: Thanks for the feedback, resolving to keep the number of active threads under control 😅

  about lifecycle transitions/state of the Cluster and the underlying components as well.

### Non-Goals/Future Work

- Resolving all the idiosyncrasies that exists in Cluster API, core Kubernetes, the rest of the ecosystem.
  (Let’s stay focused in Cluster API and keep improving incrementally).

## Proposal

This proposal groups a set of changes to status fields in Cluster API resources.

Some of those changes could be considered straight forward, e.g.

> - K8s API conventions suggest to deprecate and remove `phase` fields from status, Cluster API is going to align to this recommendation

Reviewer 4: One positive note about `phase` is the immediate feedback that it provides when running things like `kubectl get machines` or other resources. It provides a clear and simple feedback on "where are we at".

Given that this change is geared towards end users, do we agree that we need a replacement first before we deprecate something that's considered generally useful?

Author: I can see both reasons.

[K8s guidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) have a pretty good argument about dropping phase, and unless I'm mistaken, they have already dropped it from print columns in most of the core types.

But also I get it is intuitive for users.

What made me lean towards the decision to drop, is the fact that IMO phases in CAPI objects are mostly about the provisioning workflow, while they are almost useless for the rest of the lifecycle of the objects (except might be for deletion).

But happy to chat about it

Author: Considering also other discussions, we are keeping phases (and intentionally diverging from guidelines for this point). Also added a paragraph explaining why we took this decision

  (and improve Conditions to provide similar or even a better info as a replacement).
- K8s resources do not have a concept similar to "terminal feature" existing in Cluster API resources, and users approaching
  the project are struggling with this idea; in some cases also provider's implementers are struggling with it.
  Accordingly, Cluster API resources are dropping `FailureReason` and `FailureMessage` fields (terminal failures should be surfaced using

>   conditions, like any other error/warning/message)

Reviewer 4: Agreed on this one, the terminal failures today at least are often "solved" through MachineHealthCheck in some ways.

Author: Thanks for the feedback, resolving to keep the number of active threads under control 😅

> - Bubble up more information about both CP and worker Machines to the Cluster level.

Reviewer 5: Not sure it needs to be mentioned explicitly but I would also include Bubble up more information about Machines to the owner scalable resources. e.g. when I look at a MachineDeployment I'd like to have an understanding of why my Machines are failing. This would require have a clear contract to bubble up errors all the way from infraMachines.

Reviewer 5: Actually may be worth having a User stories section for things like this

Reviewer 4: Agreed on this one, as a user, MachineSet and MachineDeployment should surface as much information as possible when things aren't working

Author: Clarified the sentence and also the corresponding goal.
The implementation was already there (and BTW, I fully agree)

Author: Resolving old comments without feedback, but it should be fixed.

Some other changes requires a little bit more context, which is provided in following paragraphs:

- Review and standardize the usage of the concept of readiness and availability to align to K8s API conventions / 
  conditions used in core K8s objects like `Pod`, `Node`, `Deployment`, `ReplicaSet` etc. 
- Transition to K8s API conventions fully aligned conditions types/condition management (and thus deprecation of
  the Cluster API "custom" guidelines for conditions).

The last set of changes is a consequence of the above changes, or small improvements to address feedback received
over time; changes in this group will be detailed case by case in the following paragraphs, a few examples:

- Change the semantic of ReadyReplica counters to use Machine's Ready condition instead of Node's Ready condition.

>   (so everywhere Ready is used for a Machine it always means the same thing)

Reviewer 4: Machine is ready if and only if the Node is Ready, that definition is still in place with this change?

Author: Yes, machine ready includes NodeHealthy, which is NodeReady and not (MemoryPressure, DiskPressure etc)

Author: Hopefully this is clear now, closing this thread

> - Add a new condition monitoring the status of the connectivity to workload clusters (`RemoteConnectionProbe`).

Reviewer 4: Amazing

Author: Thanks for the feedback, I agree this is a very welcome improvement! 
Resolving this to keep the number of open threads under control 😅

Other resources will be added as soon as there will be agreement on the general direction.

Overall, the union of all those changes, is expected to greatly improve status fields, conditions, replica counters 
and print columns.

Those improvements are expected to provide benefit to users interacting with the systems, using monitoring tools, and 
building higher level systems or products on top of Cluster API.

### Readiness and Availability

The [condition CAEP](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20200506-conditions.md) in Cluster API introduced very strict requirements about “Ready” condition, mandating it 
to exists on all resources and also mandating that Ready must be computed as the summary of all the to other existing
conditions.

However, over time Cluster API maintainers recognized several limitations of the “one fit all”, strict approach.

e.g. Higher level abstractions in Cluster API are designed to remain operational during lifecycle operations,
for instance a Machine deployment is operational even if is rolling out.

But the use cases above where hard to combine with the strict requirement to have all the conditions true, and

> as a result today Cluster APi resources barely have conditions surfacing that lifecycle operations are happening, or where

Reviewer 5: typo API

> those condition are defined they have a semantic which is not easy to understand, like e.g. 'Resized' or 'MachinesSpecUpToDate'.

Reviewer 5: typo conditions

e.g. when you look at higher level abstractions in Cluster API like Clusters, MachineDeployments and ControlPlanes, readiness
might be confusing, because those resources usually accept a certain degree of not readiness, e.g. MachineDeployments are

> usually ok even if a few machine is not ready (up to MaxUnavailable).

Reviewer 4: This is a great point, and thank you for calling it out. More often than not when we look for "Ready" we're looking for things to be healthy. Which might be better suited for a different signal.

Author: Thanks for the feedback, resolving to keep the number of open threads under control

> In order to address thi problem, Cluster API is going to align to K8s API conventions. As a consequence, the “Ready”

Reviewer 5: typo this

### Transition to K8s API conventions aligned conditions

K8s is undergoing an effort of standardizing usage of conditions across all resource types, and the transition to
the v1beta2 API version is a great opportunity for Cluster API to align to this effort.

The value of this transition is substantial, because the differences that exists today's are really confusing for users;
those differences are also making it harder for ecosystem tools to build on top of Cluster API, and in some cases
even confusing new (and old) contributors.

With this proposal Cluster API will close the gap with K8s API conventions in regard to:
- Polarity: Condition type names should make sense for humans; neither positive nor negative polarity can be recommended
  as a general rule (already implemented by [#10550](https://github.com/kubernetes-sigs/cluster-api/pull/10550))
- Use of the Reason field is required (currently in Cluster API reasons is added only when condition are false)
- Controllers should apply their conditions to a resource the first time they visit the resource, even if the status is Unknown.
  (currently Cluster API controllers add conditions at different stages of the reconcile loops)

> - Cluster API is also dropping its own Condition type and start using metav1.Conditions from the Kubernetes API.

Reviewer 4: Is there a specific min Kubernetes version we're targeting?

Author: Not really, 
this depends only by the Kubernetes version we are using in CAPI go.mod file, and we are already at a version that has all the types we need to have

Reviewer 2: If I see correctly metav1.Conditions are unchanged since 4 years: https://github.com/kubernetes/kubernetes/blame/3f7d4b787b74a2d110901d333f83e99c0e5e49c9/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1532

(not sure why you're asking though :))

Author: Resolving, I assume this is clear now

The last point have also another implication, which is the removal of the Severity field which is currently used
to determine priority when merging conditions.

> TODO Document how we are going to replace severity (currently prototyping)

Reviewer 4: One example I've had in mind was something like: 
- Say I'm creating ec2 instances
- Try to create instance, but waiting for some other component (like IPAM) to allocate address
- A `InstanceCreated` condition could be false, in warning state, and explain why.

One could argue we don't really need to distinguish between "giving a warning" vs "error/failure" and that's enough for that condition to just report false; which might be ok after all?

> TODO Document how we are going to replace severity (currently prototyping)

Author: I finally had time to write this up, PTAL
TL;DR; I don't think that we will need this because with the revised concept or readyness it is probably better to surface all the messages vs arbitrarily hiding some (and merge priority to pick the single message to surface when computing ready).

But I think that our conditions utils should make it possible for developers to plug in custom func to determine merge priority for conditions, if the need this for some use case (conditions util were used broadly in providers, and now that we are aligning them to K8s, they can potentially be used in more use cases).

> TODO Document how we are going to replace severity (currently prototyping)

Reviewer 6: I like the idea of having an ability to customise the priority. If X condition and Y condition are in their non-normative state, leaving it to the provider to suggest which one is a priority is very flexible and less likely to cause issues if we make blanket assumptions, SGTM

> TODO Document how we are going to replace severity (currently prototyping)

Reviewer 2: Agree, all good as long as it is a "local" decision by the provider (or in general the controller owning the conditions / resource). I think we should just not try to "hand over" that priority information to the core CAPI controller

> TODO Document how we are going to replace severity (currently prototyping)

Author: It seems there is consensus, resolving

### Changes to Machine resource

#### Machine Status

Following changes are implemented to Machine's status:

- Disambiguate usage of ready term by renaming fields used for the provisioning workflow
- Align to K8s API conventions by deprecating `Phase` and corresponding `LastUpdated`

> - Remove `FailureReason` and `FailureMessage` to get rid of the confusing concept of terminal failures

Reviewer 5: may be specify is up to a consumer considering a condition type/reason a "terminal failure"?

Author: done, PTAL

Below you can find the relevant fields in Machine Status v1beta2, after v1beta1 removal (end state);
After golang types, you can find a summary table that also shows how changes will be rolled out according to K8s deprecation rules.

```golang
type MachineStatus struct {

    // Initialization provides observations of the Machine initialization process.
    // NOTE: fields in this struct are part of the Cluster API contract and are used to orchestrate initial Machine provisioning.
    // The value of those fields is never updated after provisioning is completed.
    // Use conditions to monitor the operational state of the Machine's BootstrapSecret.
    // +optional
    Initialization *MachineInitializationStatus `json:"initialization,omitempty"`
    
    // Represents the observations of a Machine's current state.
    // +listType=map
    // +listMapKey=type
    Conditions []metav1.Condition `json:"conditions,omitempty"`
    
    // Other fields...
    // NOTE: `Phase`, `LastUpdated`, `FailureReason`, `FailureMessage` fields won't be there anymore
}

// MachineInitializationStatus provides observations of the Machine initialization process.

> type MachineInitializationStatus struct {

Reviewer 4: Pointed this out above, but how are we going to provide some feedback in `kubectl get machines`?

Author: this will surface in print columns because the machine won't be ready (pending the discussion on status)

Author: resolving because I'm consolidating all the discussion about status.phase into https://github.com/kubernetes-sigs/cluster-api/pull/10897#discussion_r1685919394

    // BootstrapSecretCreated is true when the bootstrap provider reports that the Machine's boostrap secret is created.
    // NOTE: this field is part of the Cluster API contract, and it is used to orchestrate initial Machine provisioning.
    // The value of this field is never updated after provisioning is completed.
    // Use conditions to monitor the operational state of the Machine's BootstrapSecret.
    // +optional

>     BootstrapSecretCreated bool `json:"bootstrapSecretCreated"`

Reviewer 4: Isn't that hte same as having a phase field though?

>     BootstrapSecretCreated bool `json:"bootstrapSecretCreated"`

Author: > Have you considered using an enum for this instead, perhaps it would be better to represent this in a way that allows more states.

This is just a flag to signal signal progress during the inital provisioning.
There will be a `BootstrapConfigReady` condition to surface what is happening before and after BootstrapSecretCreated.

(So the answer is yes, it is just addressed using conditions, like other things having more states)

>     BootstrapSecretCreated bool `json:"bootstrapSecretCreated"`

Reviewer 6: > Isn't that hte same as having a phase field though?

Right yes, I think it is

> This is just a flag to signal signal progress during the inital provisioning.

If you have the condition, and the condition eventually reaches some terminal happy state, and reaching that terminal happy state flicks this bool to true, do we really need the bool, or should the consumer of this field just check the condition status?

>     BootstrapSecretCreated bool `json:"bootstrapSecretCreated"`

Author: > The value of this field is never updated after provisioning is completed.
   Use conditions to monitor the operational state of the Machine's BootstrapSecret.
   
The condition can flick after initial provisioning is completed (the condition doesn't have a "terminal" happy state).
e.g. After initial provisioning of the secret, the condition goes to false if the controller fails to renew the token in the workload cluster.

>     BootstrapSecretCreated bool `json:"bootstrapSecretCreated"`

Reviewer 6: If the condition flicks to false, why would this field not also flick to false? Is the semantic of the field here that the secret was once available rather than is currently available? What use case are we solving by having that separate?

>     BootstrapSecretCreated bool `json:"bootstrapSecretCreated"`

Reviewer 2: I think we're coming from that we want to continue to support the "initial provisioning / bootstraping" workflow of Machines. We are only using the BootstrapSecret *once* for a **Machine** during initial bootstrap.

From the other side we see value in having continuous Ready state for the InfraMachine objects, so that becomes then the question if we should do the same for BootstrapConfig. Also there are cases where BootstrapConfigs are used more than once (at least MachinePools).

On the other side the `BootstrapConfigReady` condition can give us precise error information in case the condition doesn't become ready, which is not really possible with a bool or enum field.

>     BootstrapSecretCreated bool `json:"bootstrapSecretCreated"`

Author: FYI I have added a new non-goal/paragraph about the impacts this proposal on the Cluster API contract; hopefully this could also help to address this discussion (changes are in a separated commit for better review)

TL;DR
This proposal is not going to change _how_ the Cluster API contract with infrastructure, bootstrap and control providers currently works (by using status fields like `status.ready`, bool).

However, this proposal going to change a few field names in the Cluster API contract with the only goal _to disambiguate the usage of the ready term_ by renaming fields used for the initial provisioning workflow (e.g. from `status.ready` to `status.initialization.dataSecretCreated`, which surfaces as `status.initialization.bootstapDataSecretCreated`).

While doing this, we also try to better document what is the current behaviour of Cluster API contract fields.

Changing the current behaviour of Cluster API contract fields, e.g allow those fields to flick value after provisioning, or replace them with conditions is out of scope of this proposal (note: I'm not saying it is a good or bad thing, I'm just trying to keep the scope of this work under control 😅)

    
    // InfrastructureProvisioned is true when the infrastructure provider reports that the Machine's infrastructure is fully provisioned.
    // NOTE: this field is part of the Cluster API contract, and it is used to orchestrate initial Machine  provisioning.
    // The value of this field is never updated after provisioning is completed.
    // Use conditions to monitor the operational state of the Machine's infrastructure.
    // +optional

>     InfrastructureProvisioned bool `json:"infrastructureProvisioned"`

Author: same
(for the initial the initial provisioning we need just true/false, for a more detailed reporting of what happens before and after provisioning it is already addressed using conditions, like other things having more states)

}
```

> | v1beta1 (current)             | v1beta2 (tentative Q1 2025)                              | v1beta1 removal (tentative Q1 2026)        |

Reviewer 4: Is the third column actually v1beta3?

Author: based on  https://kubernetes.io/docs/reference/using-api/deprecation-policy/#fields-of-rest-resources, it will be possible to drop back compatibily columns when we stop serving v1beta1, but please double check my assumption.

Reviewer 2: While it also was initially surprising to me the language there seems to be pretty clear
> When v1 is eventually removed, the deprecated field can be removed from v2.

(and the example is even with GA apiVersions not even just beta)

Author: Resolving, it seems that there is agreement on how to read the API guidelines

|                               | `BackCompatibilty` (new)                                 | (removed)                                  |
| `Phase` (deprecated)          | `BackCompatibilty.Phase` (renamed) (deprecated)          | (removed)                                  |
| `LastUpdated` (deprecated)    | `BackCompatibilty.LastUpdated` (renamed) (deprecated)    | (removed)                                  |
| `FailureReason` (deprecated)  | `BackCompatibilty.FailureReason` (renamed) (deprecated)  | (removed)                                  |
| `FailureMessage` (deprecated) | `BackCompatibilty.FailureMessage` (renamed) (deprecated) | (removed)                                  |
| `Conditions`                  | `BackCompatibilty.Conditions` (renamed) (deprecated)     | (removed)                                  |
| `ExprimentalConditions` (new) | `Conditions` (renamed)                                   | `Conditions`                               |
| other fields...               | other fields...                                          | other fields...                            |

Notes:

> - The `BackCompatibilty` struct is going to exist in v1beta2 types only until v1beta1 removal (9 months or 3 minor releases after v1beta2 is released/v1beta1 is deprecated, whichever is longer).

Reviewer 4: Why do we need to keep this in a sub-struct in v1beta2?

> - The `BackCompatibilty` struct is going to exist in v1beta2 types only until v1beta1 removal (9 months or 3 minor releases after v1beta2 is released/v1beta1 is deprecated, whichever is longer).

Author: I took this idea from https://kubernetes.io/docs/reference/using-api/deprecation-policy/#fields-of-rest-resources.

This was sort of turning point in preparing this design proposal, because the fact of having two fields (one representing the end state, one for back compatibility / down conversions) is what allowed me to think about an ideal target state.

> - The `BackCompatibilty` struct is going to exist in v1beta2 types only until v1beta1 removal (9 months or 3 minor releases after v1beta2 is released/v1beta1 is deprecated, whichever is longer).

Reviewer 6: I agree with the interpretation there, and I think I can answer my own question, yes, this would be set by the controller. SGTM on this concept at least, I think I had a question about the name but we can discuss on the other thread

> - The `BackCompatibilty` struct is going to exist in v1beta2 types only until v1beta1 removal (9 months or 3 minor releases after v1beta2 is released/v1beta1 is deprecated, whichever is longer).

Reviewer 2: Agree. I think we have to ensure that folks using v1beta1 get conditions with the same semantic as today. This is not possible with down conversion from v1beta2 as some of the new conditions have a different semantic

> - The `BackCompatibilty` struct is going to exist in v1beta2 types only until v1beta1 removal (9 months or 3 minor releases after v1beta2 is released/v1beta1 is deprecated, whichever is longer).

Author: Resolving this comment since there is agreement.
Also, there are other threads discussion the naming for the back compatibility field, so they willl be still in the radar

| Condition              | Note                                                                                                                                                                                                                                                                        |
|------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `Available`            | True if at the machine is Ready for at least MinReady seconds, as defined by the Machine's owner resource                                                                                                                                                                   |
| `Ready`                | True if Machine's `BootstrapSecretReady`, `InfrastructureReady`, `NodeHealthy` and `HealthCheckSucceeded` (if present) are true; if other conditions are defined in `spec.readinessGates` are defined, those conditions should be true as well for the Machine to be ready. |

> | `UpToDate`             | True if the Machine spec matches the spec of the Machine's owner resource, e.g KubeadmControlPlane or MachineDeployment                                                                                                                                                     |

Reviewer 4: In which cases the Machine spec wouldn't match and who's gonna do the checking?

Author: The idea is that it will be up to the Machine's owner controllers to set this condition.

E.g. With a machine deployment, after changing spec, all the existing machines - under a now old MachineSet - will get UpToDate=False, while the replacement machines created under the new current MachineSet will be created with UpToDate=True.

Also, this is something that opens up to track in place updates we are discussing in the working group

Reviewer 2: \* I think the source of truth for the UpToDate condition has to come from the owning controller (whoever that might be). Not sure if that controller should also directly write the condition.

For example we could also just let the owning controller mark the resource as outdated and then the Machine controller writes the condition. We already ~ use this pattern with the revision annotations. Just might be better to have a clear standardized way to signal that a Machine is outdated. This would also provide a clear API surface for folks to signal a Machine is outdated, e.g. for other control plane providers. We shouldn't have some sort of "internal" API with KCP.

Reviewer 2: @fabriziopandini IIRC the idea going forward was also that UpToDate is false if an in-place upgrade was pending. I think it would be really good if the Machine controller is writing this condition. It feels like a mix of responsibilities if the owning controller has to figure out if an in-place upgrade is (still) ongoing for a Machine

Author: Resolving this thread, hopefully this is clear now

#### Machine Spec

Machine's spec is going to be improved to allow 3rd party to extend the semantic of the new Machine's `Ready` condition
as well to standardize the concept of Machine's `Availability`.

Below you can find the relevant fields in Machine Status v1beta2, after v1beta1 removal (end state);
After golang types, you can find a summary table that also shows how changes will be rolled out according to K8s deprecation rules.

```go
type MachineSpec struct {
    
    // MinReadySeconds is the minimum number of seconds for which a Node for a newly created machine should be ready before considering the replica available.
    // Defaults to 0 (machine will be considered available as soon as the Node is ready)
    // +optional
    MinReadySeconds int32 `json:"minReadySeconds,omitempty"`

    // If specified, all readiness gates will be evaluated for Machine readiness.
    // A Machine is ready when `InfrastructureReady`, `NodeHealthy` and `HealthCheckSucceeded` (if present) are "True"; 
    // if other conditions are defined this field, those conditions should be "True" as well for the Machine to be ready.
    // +optional
    // +listType=map
    // +listMapKey=conditionType

>     ReadinessGates []MachineReadinessGate `json:"readinessGates,omitempty"`

Reviewer 4: Kind of agree, it doesn't feel this field really belongs within the Machine itself but could be a contract with Infrastructure and Bootstrap provider to expose that information in status?

>     ReadinessGates []MachineReadinessGate `json:"readinessGates,omitempty"`

Author: If I look at prior art, also pods have rediness gates in their spec

Also, it kind makes sense to me when I think to e.g. KCP creating machines and adding things like APIServerPodHealthy to the readiness gate in a simple and straight forward way (if instead KCP has to bubble up the same info in status, then CAPI has to apply it to control plane machines after they are created by KCP, which seems counter-intuitive).

Similarly, I can see external controllers, e.g. responsible to install special software/hardware on the machines doing the same even if they are not actual CAPI controllers

>     ReadinessGates []MachineReadinessGate `json:"readinessGates,omitempty"`

Reviewer 6: Ack, that seems like a reasonable use case, hadn't thought about the controllers adding these in.

Can we make sure this is clarified as some of the expected use cases somewhere within the doc? This is useful context I wouldn't want to lose to GitHub threads

I would also suggest adding some wording to the godoc for the field that suggests similar

>     ReadinessGates []MachineReadinessGate `json:"readinessGates,omitempty"`

Author: Resolving given that it seems that there is agreement.
(I have also added the examples to the godoc)

    // Other fields...
}

// MachineReadinessGate contains the reference to a Machine condition to be used as readiness gates.
type MachineReadinessGate struct {
    // ConditionType refers to a condition in the Machine's condition list with matching type.
    // Note: Both  Cluster API conditions or conditions added by 3rd party controller can be used as readiness gates.
    ConditionType string `json:"conditionType"`
}
```

| v1beta1 (current)      | v1Beta2 (tentative Q1 2025) | v1beta1 removal (tentative Q1 2026) |
|------------------------|-----------------------------|-------------------------------------|
| `ReadinessGates` (new) | `ReadinessGates`            | `ReadinessGates`                    |
| other fields...        | other fields...             | other fields...                     |

Notes:
- Both `MinReadySeconds` and `ReadinessGates` should be treated as other in-place propagated fields (changing this should not trigger rollouts).

> - Similarly to Pod's `ReadinessGates`, also Machine's `ReadinessGates` accept only conditions with positive polarity; 

Reviewer 5: Any specific reason for this? I think this might result in unnatural choices for consumers at the time of defining their readiness gates and seems a bit miss aligned with the spirit of the proposal about relaxing custom impositions and standardizing on kube. As stated in kube recommendations It's ok for negative polarity to represent normal operation

Reviewer 4: How would we validate that's only positive polarity?

Author: > Any specific reason for this? 

Mostly for sake of keeping this proposal scope under control + I don't have concrete examples that make me think we now need something more complex of what exists in K8s

But I agree with the fact that this is one of the idiosyncrasies in K8s, and so I also added in the next sentence "The Cluster API project might revisit this in the future to stay aligned with Kubernetes or if there are use cases justifying this change."

I also have a few ideas about how to improve this, but if it is not a blocker, I think we should have a minimal feature of this feature in ASAP, and then iterate based on needs.

> How would we validate that's only positive polarity?

We do not validate, it is just what we will assume when computing ready

Reviewer 2: I'm fine with only supporting aggregating positive polarity conditions into Ready for now. It should be relatively trivial to extend the readinessGates field later to also support negative polarity

Reviewer 6: readinessGates should be a slice of some small struct if we see a want to expand it in the future to indicate a desired happy state polarity

Reviewer 2: Oh yup :). That's why we propose
```go
// MachineReadinessGate contains the reference to a Machine condition to be used as readiness gates.
type MachineReadinessGate struct {
    // ConditionType refers to a condition in the Machine's condition list with matching type.
    // Note: Both  Cluster API conditions or conditions added by 3rd party controller can be used as readiness gates.
    ConditionType string `json:"conditionType"`
}
```
(which is what is k/k is doing for readinessGates as well)

Reviewer 2: @zjs I assume you mean `Status ConditionStatus`? (same field name as in the condition type)

Author: Personally I would prefer to get a first version of this feature implemented without support for negative polarity and then iterate (the API is easily extensible for this use case). Also others agree above.

But I don't have strong opinions, so I would wait for some more feedback. Please, add 👍 if you want to add support for negative polarity now, 👎 if it is ok for you to aim for a first iteration without supporting negative polarity and then follow-up.

In case we implement support for negative polarity, let's keep in mind https://github.com/kubernetes-sigs/cluster-api/pull/10897/files#r1692281748

Reviewer 2: > aim for a first iteration without supporting negative polarity and then follow-up.

@fabriziopandini Probably good time to start an umbrella issue (where we can also have a section with backlog / follow-ups)

Author: Tracked in https://github.com/kubernetes-sigs/cluster-api/issues/11105, also the proposal already has note about the intent to consider implementing support for negative polarity conditions to ReadinessGates and AvailabilityGates

  The Cluster API project might revisit this in future to stay aligned with Kubernetes or if there are use cases justifying this change.

#### Machine Print columns

> | Current           | To be                         |

Author: It is already there 😉 
All the the columns displayed only with -o wide are marked with (*) - linking to a corresponding foot note

> | Current           | To be                         |

Reviewer 6: I clearly missed that point 😅 Thanks

|-------------------|-------------------------------|
| `NAME`            | `NAME`                        |
| `CLUSTER`         | `CLUSTER`                     |
| `NODE NAME`       | `PAUSED` (new) (*)            |
| `PROVIDER ID`     | `NODE NAME`                   |

> | `PHASE` (deleted) | `PROVIDER ID`                 |

Reviewer 4: Going to sound like a broken record, apologies :D ! One of the goals is to improve the user friendliness of internals. I _feel_ a bit though that the current `phase` field communicates better that an explosion of fields

Author: Let's chat about this, I'm sure we can find our way to what matters for users.

Reviewer 7: + 1 on "this is FANTASTIC and AMAZING work". Thanks @fabriziopandini.
also argee on that the status.ready(bool type) and status.phase(enum type) communicate better than a set of fields or conditions. Although if it can be calculated by the high level consumer or internal code of Cluster API, what if Cluster API object controller calculates it and record in status.ready and status.phase ?
btw @sbueringer i think status.phase is used to show the current provisioning or operation state, e.g. Creating/Upgrading/ScalingUp etc.

Reviewer 2: > btw @sbueringer i think status.phase is used to show the current provisioning or operation state, e.g. Creating/Upgrading/ScalingUp etc.

Depends on the object. I was here referring to the Machine object (because this table here is about the Machine object). MachineDeployment phase has also ScalingUp & ScaldingDown.

Author: Resolving because I'm consolidating all the discussions about status.phases into https://github.com/kubernetes-sigs/cluster-api/pull/10897#discussion_r1685919394

| `AGE`             | `READY` (new)                 |
| `VERSION`         | `AVAILABLE` (new)             |
|                   | `UP TO DATE` (new)            |
|                   | `AGE`                         |
|                   | `OS-IMAGE` (new) (*)          |
|                   | `KERNEL-VERSION` (new) (*)    |
|                   | `CONTAINER-RUNTIME` (new) (*) |

> TODO: figure out if can `INTERNAL-IP` (new) (*),   `EXTERNAL-IP` after `VERSION` / before `OS-IMAGE`?  (similar to Nodes...).

Reviewer 4: +++1 and would love if we can get all of these in -o wide

Author: Thanks for the feedback! 💯  agree!

Resolving to keep the number of open threads under control 😅

| `ReadyReplicas` (deprecated)     | `BackCompatibilty.ReadyReplicas` (renamed) (deprecated)  | (removed)                           |
| `ExprimentalReadyReplicas` (new) | `ReadyReplicas` (renamed)                                | `ReadyReplicas`                     |
| `FailureReason` (deprecated)     | `BackCompatibilty.FailureReason` (renamed) (deprecated)  | (removed)                           |
| `FailureMessage` (deprecated)    | `BackCompatibilty.FailureMessage` (renamed) (deprecated) | (removed)                           |
| `Conditions`                     | `BackCompatibilty.Conditions` (renamed) (deprecated)     | (removed)                           |
| `ExprimentalConditions` (new)    | `Conditions` (renamed)                                   | `Conditions`                        |
| `UpToDateReplicas` (new)         | `UpToDateReplicas`                                       | `UpToDateReplicas`                  |
| other fields...                  | other fields...                                          | other fields...                     |

Notes:

> - The `BackCompatibilty` struct is going to exist in v1beta2 types only until v1beta1 removal (9 months or 3 minor releases after v1beta2 is released/v1beta1 is deprecated, whichever is longer).

Reviewer 4: We could just not add any backward compatibility here and just try to support conversions instead?

> - The `BackCompatibilty` struct is going to exist in v1beta2 types only until v1beta1 removal (9 months or 3 minor releases after v1beta2 is released/v1beta1 is deprecated, whichever is longer).

Author: I took this idea from https://kubernetes.io/docs/reference/using-api/deprecation-policy/#fields-of-rest-resources.

I'm not opposed to other names (`v1beta1`, `BackCompatibility.v1beta1`, `Deprecated` etc), happy to take suggestions.

But IMO the real value of using this pattern for this transition is that allows us to design the ideal state, without caring about down conversions, which in my opinion could be tricky to implement (and frankly speaking not sure if worth the effort vs keeping old fields around for some more time).

> - The `BackCompatibilty` struct is going to exist in v1beta2 types only until v1beta1 removal (9 months or 3 minor releases after v1beta2 is released/v1beta1 is deprecated, whichever is longer).

Reviewer 2: I think down conversion from v1beta2 conditions to v1beta1 conditions is only possible if the semantic is basically identical, because we can't really gather additional information in a conversion webhook. 

So I think it's the better option to have both fields side-by-side. It also makes it possible for folks to gradually migrate because they can access both the old and new conditions with both the v1beta1 and v1beta2 API types.

> - The `BackCompatibilty` struct is going to exist in v1beta2 types only until v1beta1 removal (9 months or 3 minor releases after v1beta2 is released/v1beta1 is deprecated, whichever is longer).

Reviewer 2: > Were other names considered for the struct, would deprecated be a suitable alternative? 

Regarding naming. I would suggest `deprecatedV1beta1.conditions`, ...

Similarly I would suggest `v1beta2Conditions` or `v1beta2.conditions` (also for other fields like `v1beta2.readyReplicas` instead of `ExperimentalConditions` / `ExperimentalReadyReplicas`

> - The `BackCompatibilty` struct is going to exist in v1beta2 types only until v1beta1 removal (9 months or 3 minor releases after v1beta2 is released/v1beta1 is deprecated, whichever is longer).

Author: I'm fine with any of the suggested names (my preference goes to `V1beta2.*` and `V1beta1.Deprecated.*` for consistency) but I will wait for a few more +1 before changing this

> - The `BackCompatibilty` struct is going to exist in v1beta2 types only until v1beta1 removal (9 months or 3 minor releases after v1beta2 is released/v1beta1 is deprecated, whichever is longer).

Reviewer 2: Fine for me!

> - The `BackCompatibilty` struct is going to exist in v1beta2 types only until v1beta1 removal (9 months or 3 minor releases after v1beta2 is released/v1beta1 is deprecated, whichever is longer).

Author: Leaning towards

```yaml
    # phase1
        status:
          conditions: ...
          readyReplicas: ...
          v1beta2: # new fields under a struct, not using "experimental"
              readyReplicas: ...
              conditions: ...          
    
    # phase2
      status:
        deprecated: # << old fields preserved for down conversion and moved under a deprecated struct
          v1beta1:  # << additional level of nesting to make it extensible
              readyReplicas: ...
              conditions: ...
        readyReplicas: ...
        conditions: ...
    
    # phase3
      status:
        readyReplicas: ...
        conditions: ...
```

  Fields in this struct are used for supporting down conversions, thus providing users relying on v1beta1 APIs additional buffer time to pick up the new changes.
- This proposal is using `UpToDateReplicas` instead of `UpdatedReplicas`; This is a deliberated choice to avoid 
  confusion between update (any change) and upgrade (change of the Kubernetes versions).
- Also `AvailableReplicas` will determine Machine's availability by reading Machine.Available condition instead of
  computing availability as of today, however in this case the semantic of the field is not changed

TODO: check `FullyLabeledReplicas`, do we still need it?

#### MachineSet (New)Conditions

| Condition        | Note                                                                                                           |
|------------------|----------------------------------------------------------------------------------------------------------------|

> | `ReplicaFailure` | This condition surfaces issues on creating Machines, if any.                                                   |

Reviewer 5: It's not obvious to me the diff between ReplicaFailure and MachinesReady

Author: Improved the description, PTAL
FYI The idea comes from NewReplicaSetAvailable in Kubernetes, but I don't feel strongly about this condition and we might also drop it entirely

| `ScalingUp`      | True if available replicas < desired replicas                                                                  |
| `ScalingDown`    | True if replicas > desired replicas                                                                            |
| `UpToDate`       | True if all the Machines controlled by this MachineSet are up to date (replicas = upToDate replicas)           |
| `Remediating`    | True if there is at least one Machine controlled by this MachineSet is not passing health checks               |
| `Deleted`        | True if MachineSet is deleted; Reason can be used to observe the cleanup progress when the resource is deleted |
| `Paused`         | True if this MachineSet or the Cluster it belongs to are paused                                                |

> To better evaluate proposed changes, below you can find the list of current MachineSet's conditions:
> Ready, MachinesCreated, Resized, MachinesReady.

Notes:
- MachineSet conditions are intentionally mostly consistent with MachineDeployment conditions to help users troubleshooting .

> - MachineSet is considered as a sort of implementation detail of MachineDeployments, so it doesn't have its own concept of availability.

Author: If I look at prior art, the same assumption was used for ReplicaSet in K8s.

I think this is also the case for CAPI, because also MachineSet do not have any sort of "not aivaliabiltiy" they tolerate, and thus it does not make sense to talk about availabity too

  Similarly, this proposal is dropping the notion of MachineSet readiness because it is preferred to let users focusing on Machines readiness.
- `Remediating` for older MachineSet sets will report that remediation will happen as part of the regular rollout.
- `UpToDate` condition initially will be `false` for older MachineSet, `true` for the current MachineSet; however in
  the future the latter might evolve in case Cluster API will start supporting in-place upgrades.

#### MachineSet Print columns

| Current       | To be                   |
|---------------|-------------------------|
| `NAME`        | `NAME`                  |
| `CLUSTER`     | `CLUSTER`               |
| `DESIRED` (*) | `PAUSED` (new) (*)      |
| `REPLICAS`    | `DESIRED`               |
| `READY`       | `CURRENT` (renamed) (*) |
| `AVAILABLE`   | `READY` (updated)       |
| `AGE`         | `AVAILABLE` (updated)   |
| `VERSION`     | `UP-TO-DATE` (new)      |
|               | `AGE`                   |

> |               | `VERSION`               |

Reviewer 4: Could we maybe also add the name of the bootstrapRef and infraRef resource? or is that too much info?

Author: Ideally we can, but I think we should figure it out when we implement it and we can see how it looks like.
Also, print columns are not subject to API guarantee, so we are free to iteratively improve this anytime.

Author: Added a not about this, resolving

(also trying to keep the number of open threads under control 😅 )

| `FailureMessage` (deprecated) | `BackCompatibilty.FailureMessage` (renamed) (deprecated) | (removed)                           |
| `Conditions`                  | `BackCompatibilty.Conditions` (renamed) (deprecated)     | (removed)                           |
| `ExprimentalConditions` (new) | `Conditions` (renamed)                                   | `Conditions`                        |
| other fields...               | other fields...                                          | other fields...                     |

Notes:
- The `BackCompatibilty` struct is going to exist in v1beta2 types only until v1beta1 removal (9 months or 3 minor releases after v1beta2 is released/v1beta1 is deprecated, whichever is longer).
  Fields in this struct are used for supporting down conversions, thus providing users relying on v1beta1 APIs additional buffer time to pick up the new changes.

#### MachineDeployment (New)Conditions

| Condition        | Note                                                                                                                                                                                                                                                   |

> |------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

Reviewer 5: From a consumer pov it'd be valuable to look at a condition and understand if may MachineDeployment is performing a rolling upgrade and it was triggered by an infraMachineTemplate change or a kube version change

Author: Agreed, we already did some work to surface the reason why a rollout happens il logs, the next logical step is to have this in conditions too, but there is still some work to go from the raw diff we have today to a simple summary like the one you suggested.

Author: Resolving old comments without feedback.
Also added as a point in https://github.com/kubernetes-sigs/cluster-api/issues/11105

| `Available`      | True if the MachineDeployment has minimum availability according to parameters specified in the deployment strategy, e.g. If using RollingUpgrade strategy, availableReplicas must be greater or equal than desired replicas - MaxUnavailable replicas |  

> | `ReplicaFailure` | This condition surfaces issues on creating Machines controlled by this MachineDeployment, if any.                                                                                                                                                      |

Reviewer 4: Does this bubble up all the failure messages as well? It'd be great if so

Author: Clarified, also answered to a similar comment from @enxebre, PTAL

`ReplicaFailure` is just about the creation of the Machine replica in K8s.
What you are looking for (failure messages from all the machines) will be `MachinesReady`, and it will bubble up messages for the entire lifecycle of the machine, not only for creation.

Author: I have dropped ReplicaFailure, this will surface in MachinesReady (or MachinesAvailable). Name TBD

> 

Reviewer 2: nit: we should add the date as prefix to the doc (not sure if breaks the open conversations though)

      - [MachineDeployment Status](#machinedeployment-status)
      - [MachineDeployment (New)Conditions](#machinedeployment-newconditions)
      - [MachineDeployment Print columns](#machinedeployment-print-columns)
    - [Changes to Cluster resource](#changes-to-cluster-resource)
      - [Cluster Status](#cluster-status)
        - [Cluster (New)Conditions](#cluster-newconditions)
      - [Cluster Spec](#cluster-spec)
      - [Cluster Print columns](#cluster-print-columns)
    - [Changes to KubeadmControlPlane (KCP) resource](#changes-to-kubeadmcontrolplane-kcp-resource)
      - [KubeadmControlPlane Status](#kubeadmcontrolplane-status)
      - [KubeadmControlPlane (New)Conditions](#kubeadmcontrolplane-newconditions)
      - [KubeadmControlPlane Print columns](#kubeadmcontrolplane-print-columns)
    - [Changes to MachinePool resource](#changes-to-machinepool-resource)
    - [Changes to Cluster API contract](#changes-to-cluster-api-contract)

# Summary

This document defines how status across all Cluster API resources is going to evolve in the v1beta2 API version, focusing on
improving usability and consistency across different resources in CAPI and with the rest of the ecosystem.

> # Motivation

Reviewer 2: I think one example of this could be that eventually something like AvailableGates should be also configurable via the ClusterClass. So that "end users" don't have to worry about how Availability should be computed for a certain class of Clusters

> # Motivation

Author: Added a goal for this use case, PTAL
Also please note that there is already a note about the fact that in future we might consider extending CC to allow the definition of AvailabilityGates (first we need to get them in place 😅 )

> # Motivation

Author: Also added to https://github.com/kubernetes-sigs/cluster-api/issues/11105 for tracking

### Transition to K8s API conventions aligned conditions

K8s is undergoing an effort of standardizing usage of conditions across all resource types, and the transition to
the v1beta2 API version is a great opportunity for Cluster API to align to this effort.

The value of this transition is substantial, because the differences that exists today's are really confusing for users;
those differences are also making it harder for ecosystem tools to build on top of Cluster API, and in some cases
even confusing new (and old) contributors.

With this proposal Cluster API will close the gap with K8s API conventions in regard to:
- Polarity: Condition type names should make sense for humans; neither positive nor negative polarity can be recommended
  as a general rule (already implemented by [#10550](https://github.com/kubernetes-sigs/cluster-api/pull/10550))
- Use of the `Reason` field is required (currently in Cluster API reasons is added only when condition are false)
- Controllers should apply their conditions to a resource the first time they visit the resource, even if the status is `Unknown`.

>   (currently Cluster API controllers add conditions at different stages of the reconcile loops)

Reviewer 2: I think we'll have to make well thought through exceptions of this general rule. HealthCheckSucceeded should only be on Machines that are actually monitored by an MHC. control plane pod conditions should be only control plane machines, ...

>   (currently Cluster API controllers add conditions at different stages of the reconcile loops)

Author: I have expanded this point documenting that conditions applied by different controllers might show up at different time, and also adding a note from Kubernetes API conventions about how to treat missing conditions (and yes, Kubernetes API conventions already account for exceptions from this rule).

PTAL

    // The value of those fields is never updated after provisioning is completed.
    // Use conditions to monitor the operational state of the Machine.
    // +optional
    Initialization *MachineInitializationStatus `json:"initialization,omitempty"`
    
    // Conditions represent the observations of a Machine's current state.
    // +optional
    // +listType=map
    // +listMapKey=type
    Conditions []metav1.Condition `json:"conditions,omitempty"`
    
    // Other fields...
    // NOTE: `Phase`, `LastUpdated`, `FailureReason`, `FailureMessage`, `BootstrapReady`, `InfrastructureReady` fields won't be there anymore
}

// MachineInitializationStatus provides observations of the Machine initialization process.
type MachineInitializationStatus struct {

    // BootstrapSecretCreated is true when the bootstrap provider reports that the Machine's boostrap secret is created.
    // NOTE: this field is part of the Cluster API contract, and it is used to orchestrate initial Machine provisioning.
    // The value of this field is never updated after provisioning is completed.

>     // Use conditions to monitor the operational state of the Machine's BootstrapSecret.

Reviewer 2: Nice idea!

>     // Use conditions to monitor the operational state of the Machine's BootstrapSecret.

Author: No strong opinions from my side, I will wait a little bit before applying this change (please apply 👍 if in favour or 👎 if not to signal your opinion, if any)

>     // Use conditions to monitor the operational state of the Machine's BootstrapSecret.

Reviewer 6: `BootstrapSecretCreated` is already past tense isn't it? I'm personally a strong -1 on adding `Was` in the middle. API fields should be succinct and avoid passive voice where possible, I think adding `Was` takes a step towards passive voice

>     // Use conditions to monitor the operational state of the Machine's BootstrapSecret.

Reviewer 2: > BootstrapSecretCreated is already past tense isn't it?

Good point, thx!

>     // Use conditions to monitor the operational state of the Machine's BootstrapSecret.

Author: Resolving given that there is consensus on proposed names

  Fields in this struct are used for supporting down conversions, thus providing users relying on v1beta1 APIs additional buffer time to pick up the new changes. 

##### Machine (New)Conditions

| Condition              | Note                                                                                                                                                                                                                                                            |
|------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `Available`            | True if at the machine is Ready for at least MinReady seconds, as defined by the Machine's minReadySeconds field                                                                                                                                                |

> | `Ready`                | True if Machine's `BootstrapSecretReady`, `InfrastructureReady`, `NodeHealthy` and `HealthCheckSucceeded` (if present) are true; if other conditions are defined in `spec.readinessGates`, those conditions should be true as well for the Machine to be ready. |

Author: @sbueringer, folks, opinions on making this condition to turn false as soon as the deletion process start?

Reviewer 2: I think it should stay true. Just because a deletionTimestamp is set it doesn't mean that the Machine already gets deleted (deletion hooks, drain, ...). Machines could stay around for days after the deletionTimestamp being set without even reaching the stage where the Node is cordoned (aka they are still fully usable at this point)

Author: That's a fair point.
I will take a look if this might have an impact e.g. on how MD determine scale up or down.

@sbueringer follow up question, should we consider deleteting machines available or not (I think not, it is also probably in line on how MD works today, but I have to double check)

Reviewer 2: I see deleting as orthogonal to both ready and available. As long as Machines are ready & available I think we should surface this accordingly (even if the deletion is ongoing). Once we reach the point where the Machine becomes un-ready / un-available because of deletion we would update accordingly.

I'm fine if we want e.g. to say: "as soon as we actually start draining a Machine becomes un-ready / un-available". Because at this point it can actually not be used anymore to schedule new Pods / workload

Reviewer 2: With the definition of "machine can host workloads" it would be correct to mark a Machine as not ready as soon as we cordon. WDYT?

Reviewer 2: Discussed this with Fabrizio & Christian. We now think it's okay to consider a Machine with deletionTimestamp is not ready. Even if the Machine deletion is "stuck" at the "pre-drain.delete" hook we have to assume that folks are already "cleaning up" / "terminating" the Pod in these hooks. We'll update the Ready condition accordingly

> | `UpToDate`             | True if the Machine spec matches the spec of the Machine's owner resource, e.g KubeadmControlPlane or MachineDeployment                                                                                                                                         |

Reviewer 2: Where do you see the gap between the current description and "fully patched"?

> | `UpToDate`             | True if the Machine spec matches the spec of the Machine's owner resource, e.g KubeadmControlPlane or MachineDeployment                                                                                                                                         |

Author: Naming is tricky in this case.

In this proposal I decided to align everything - conditions and replica counters - to `UpToDate`; this is also consistent with `rollingUpdate` strategy in MachineDeployments.

All of the above are related to any change that could apply to the machine (not only to the change of the Kubernetes version).

FYI, I have also considered adding an `Upgrading` condition specifically for the change of the Kubernetes version) and renaming `UpToDate` condition into `Updating`, so we would have both `Updating`and `Upgrading`, with only one or both true depending on the change being performed (and hopefully something in conditions reason/message will help to disambiguate...).

But I'm a little bit debated about this option, so it is not in the current draft 😅 

opinions?

> | `UpToDate`             | True if the Machine spec matches the spec of the Machine's owner resource, e.g KubeadmControlPlane or MachineDeployment                                                                                                                                         |

Reviewer 2: I perfer to keep it simple for now. I think UpToDate is fine. If necessary we could add Updating and/or Upgrading later

> | `UpToDate`             | True if the Machine spec matches the spec of the Machine's owner resource, e.g KubeadmControlPlane or MachineDeployment                                                                                                                                         |

Reviewer 2: Additional note. I see a difference in semantic between UpToDate and !Updating. Basically old Machines will not be UpToDate but we will also never update them. But an Updating = false condition is unable to communicate that nuance.

So I think if we want to introduce Updating / Upgrading (now or later) it's fine to have these conditions in addition to the UpToDate condition.

> | `UpToDate`             | True if the Machine spec matches the spec of the Machine's owner resource, e.g KubeadmControlPlane or MachineDeployment                                                                                                                                         |

Author: Ok to keep it simple.
Also, this combined with the idea to surface in UpToDate what is changed in the machine (infrastructure, k8s version, bootstrap) should address this point without adding more conditions

  MachineHealthCheck controller in case a MachineHealthCheck targets the machine.
- KubeadmControlPlane also adds additional conditions to Machines, but those conditions are not included in the table above
  for sake of simplicity (however they are documented in the KubeadmControlPlane paragraph).

TODO: think carefully at remote conditions becoming unknown, this could block a few operations ... 

#### Machine Spec

Machine's spec is going to be improved to allow 3rd party components to extend the semantic of the new Machine's `Ready` condition
as well as to standardize the concept of Machine's `Availability`.

Below you can find the relevant fields in Machine Status v1beta2, after v1beta1 removal (end state);
Below the Go types, you can find a summary table that also shows how changes will be rolled out according to K8s deprecation rules.

```go
type MachineSpec struct {

    // MinReadySeconds is the minimum number of seconds for which a Machine should be ready before considering the replica available.
    // Defaults to 0 (machine will be considered available as soon as the Node is ready)
    // +optional

>     MinReadySeconds int32 `json:"minReadySeconds,omitempty"`

Reviewer 2: A user wouldn't create the Machine, but the MD. But I guess the question remains the same (especially as it defaults to 0)

A good thing, it can be configured in the ClusterClass

>     MinReadySeconds int32 `json:"minReadySeconds,omitempty"`

Author: Might be what is not clearly evident above, is the fact that `MinReadySeconds` field already exists in Cluster API MachineDeployments and MachinePools, this proposal is just surfacing this field at Machine level with the goal to allow users looking at the output of `kubectl get machines` to have have the full set of info behind the availability condition. 

WRT using `readinessGates` to express "wait for something specific" instead of waiting arbitrarily for a few seconds is one of the possible use cases for `readinessGates`, but this requires users to implement a custom controller that adds a custom condition to Machine's conditions.

Having `MinReadySeconds` can help users not willing to implement a custom controller, even if I agree that just adding some delay might be racy in most cases.

Another thing to consider, and most probably also the reason why `MinReadySeconds` field exists in CAPI today, is the fact that the same field exists in Kubernetes DaemonSets, Deployments, StatefulSet. 
Having `MinReadySeconds` also in Cluster API resources might help users approaching the project

TL;DR this proposal is not adding this feature, just making its usage more explicit.
`readinessGates` provides an alternative to handle more sophistical wait strategies.

>     MinReadySeconds int32 `json:"minReadySeconds,omitempty"`

Reviewer 2: > Having MinReadySeconds can help users not willing to implement a custom controller, even if I agree that just adding some delay might be racy in most cases.

Not sure I understand how MinReadySeconds could be racy

Otherwise agree.

I think MinReadySeconds is just an easy way to ensure Available doesn't flicker if Ready is flickering (which can easily happen because of e.g. MemoryPressure condition that is included

>     MinReadySeconds int32 `json:"minReadySeconds,omitempty"`

Author: Resolving, hopefully it is clear now

    // If specified, all readiness gates will be evaluated for Machine readiness.
    // A Machine is ready when `InfrastructureReady`, `NodeHealthy` and `HealthCheckSucceeded` (if present) are "True"; 
    // if other conditions are defined in this field, those conditions should be "True" as well for the Machine to be ready.
    // +optional
    // +listType=map
    // +listMapKey=conditionType
    ReadinessGates []MachineReadinessGate `json:"readinessGates,omitempty"`

    // Other fields...
}

// MachineReadinessGate contains the reference to a Machine condition to be used as readiness gates.
type MachineReadinessGate struct {
    // ConditionType refers to a condition in the Machine's condition list with matching type.
    // Note: Both  Cluster API conditions or conditions added by 3rd party controller can be used as readiness gates.
    ConditionType string `json:"conditionType"`

> }

Reviewer 2: Interesting idea ofr the Status to be an array instead of just one value

> }

Author: merging this thread into https://github.com/kubernetes-sigs/cluster-api/pull/10897/files#r1682623197, so we have only one thread about supporting negative polarity in readiness gates

Notes:
- Both `MinReadySeconds` and `ReadinessGates` should be treated as other in-place propagated fields (changing this should not trigger rollouts).
- Similarly to Pod's `ReadinessGates`, also Machine's `ReadinessGates` accept only conditions with positive polarity; 
  The Cluster API project might revisit this in future to stay aligned with Kubernetes or if there are use cases justifying this change.

#### Machine Print columns

| Current           | To be                         |
|-------------------|-------------------------------|
| `NAME`            | `NAME`                        |
| `CLUSTER`         | `CLUSTER`                     |
| `NODE NAME`       | `PAUSED` (new) (*)            |
| `PROVIDER ID`     | `NODE NAME`                   |
| `PHASE` (deleted) | `PROVIDER ID`                 |
| `AGE`             | `READY` (new)                 |
| `VERSION`         | `AVAILABLE` (new)             |
|                   | `UP TO DATE` (new)            |
|                   | `AGE`                         |
|                   | `OS-IMAGE` (new) (*)          |
|                   | `KERNEL-VERSION` (new) (*)    |
|                   | `CONTAINER-RUNTIME` (new) (*) |

> 

Reviewer: the net addition of 3 extra new columns in the default output could potentially lead to output that frequently spans more that one line in most terminals. See current output below. Now imagine adding 3 more columns:

```
root@421a4016968eb1464e673a6b3ac7bac4 [ ~ ]# kubectl get machine -n tkg-ns tkg-cluster-ujfv-dfcqg-k6wz2
NAME                           CLUSTER            NODENAME                       PROVIDERID                                       PHASE     AGE   VERSION
tkg-cluster-ujfv-dfcqg-k6wz2   tkg-cluster-ujfv   tkg-cluster-ujfv-dfcqg-k6wz2   vsphere://421adb58-19d2-d3c1-9067-ff100888c2ce   Running   28d   v1.29.4+vmware.3-fips.1
```
Do both 'READY' and 'AVAILABLE' need to appear in the default output?

Reviewer 2: In general agree. Our idea was to test for things like this during implementation and then finalize what goes into -o wide.

Do we have a max character number we have in mind as a minimum? Then we can take this, use a "normal" example for names / node names / ... and see that it fits in

Author: I have added below all the print columns paragraphs in the proposal the following notes to surface/reinforce what @sbueringer explained above

> Notes:
> - Print columns are not subject to any deprecation rule, so it is possible to iteratively improve print columns without waiting for the next API version.
> - During the implementation we are going to verify the resulting layout and eventually make final adjustments to the column list.

Wrt to `READY` being always visible, the reason why the colums is currently visible is because you can drill down on ready replicas counters from Cluster to Machines (via KCP, MachineDeployments/MachineSet or MachinePool) by using kubectl get; ready replicas counters/ready is consistently visible everywhere.

So, I think that after the final adjustment described above, we either should have `READY`/ready replicas everywhere or drop it everywhere.

Reviewer 2: Agree!

Looking at the example above. Looks like we use a lot of space with information like providerID (and maybe also node name) that could be a lot better used by status information like ready (the ready column takes a lot less space than node name / providerID, in a lot of cases node name is identical to machine name)

Author: Resolving because we have agreement on the fact that this requires some fine tuning during/after implementation.
Also tracked in https://github.com/kubernetes-sigs/cluster-api/issues/11105

  Stefan +1 if possible

(*) visible only when using `kubectl get -o wide`

### Changes to MachineSet resource

#### MachineSet Status

Following changes are implemented to MachineSet's status:

- Update `ReadyReplicas` counter to use the same semantic Machine's `Ready` (today it is computed a Machines with Node Ready) condition and add missing `UpToDateReplicas`.
- Remove `FailureReason` and `FailureMessage` to get rid of the confusing concept of terminal failures
- Transition to new, improved, K8s API conventions aligned conditions

Below you can find the relevant fields in Machine Status v1beta2, after v1beta1 removal (end state);
Below the Go types, you can find a summary table that also shows how changes will be rolled out according to K8s deprecation rules.

```golang
type MachineSetStatus struct {

    // The number of ready replicas for this MachineSet. A machine is considered ready when Machine's Ready condition is true.
    // +optional
    ReadyReplicas int32 `json:"readyReplicas"`
	
	// The number of available replicas for this MachineSet. A machine is considered available when Machine's Available condition is true.
	// +optional

> 	AvailableReplicas int32 `json:"availableReplicas"`

Author: It is a user choice to set MinReadySeconds and thus introduce a difference between Ready and Available, I assume that depending on the rational behind the user's choice it could make sense to look at one or the another.

Personally, I think that now that we have a clean a consistent definition of readineass and availability across the board, as a project we can decide to intentionally deviate from K8s and show only availability/availability replicas on print columns, thus giving an hint on what the users should look at (Kubernetes instead shows both available and ready replicas columns for ReplicaSets and Deployments, ).

But while I'm open to consider dropping ready replicas from print columns, I defer to a separated issues/proposal the idea of removing the readyReplicas field from status of Cluster API resources, because this implies changes to CAPI controllers which are not in scope of this proposal.

Let's hear other opinions
Please add 👍 if you are in favor ready/ready replicas _from print columns only_, or add 👎 if you are in favor of being consistent with K8s (and show both ready/ready replicas and available/available replicas)

Also related to [final adjustements on print columns](https://github.com/kubernetes-sigs/cluster-api/pull/10897/files#r1693965374)

> 	AvailableReplicas int32 `json:"availableReplicas"`

Reviewer 2: I think I'm fine either way (also depends a bit on how much space we have). Are you asking to drop it entirely or move ready into `-o wide`? Should we just decide this later when we fine-tune the columns?

> and show both ready/ready replicas and available/available replicas

What do you mean with ready/ready and available/available?

> 	AvailableReplicas int32 `json:"availableReplicas"`

Author: Resolving because we have agreement on the fact that this requires some fine tuning during/after implementation (which is also well documented in the proposal)
Also tracked in https://github.com/kubernetes-sigs/cluster-api/issues/11105

    // The number of up-to-date replicas for this MachineSet. A machine is considered up-to-date when Machine's UpToDate condition is true.
    // +optional
    UpToDateReplicas int32 `json:"upToDateReplicas"`

    // Represents the observations of a MachineSet's current state.
    // +optional
    // +listType=map
    // +listMapKey=type
    Conditions []metav1.Condition `json:"conditions,omitempty"`

    // Other fields...
    // NOTE: `FailureReason`, `FailureMessage` fields won't be there anymore
}
```

| v1beta1 (current)                    | v1beta2 (tentative Q1 2025)                                 | v1beta1 removal (tentative Q1 2026) |
|--------------------------------------|-------------------------------------------------------------|-------------------------------------|

> | `ExprimentalReadyReplicas` (new)     | `ReadyReplicas` (renamed)                                   | `ReadyReplicas`                     |

Reviewer: typo: 'Exprimental...'

| `ExprimentalAvailableReplicas` (new) | `AvailableReplicas` (renamed)                               | `AvailableReplicas`                 |
|                                      | `BackCompatibilty` (new)                                    | (removed)                           |
| `ReadyReplicas` (deprecated)         | `BackCompatibilty.ReadyReplicas` (renamed) (deprecated)     | (removed)                           |
| `AvailableReplicas` (deprecated)     | `BackCompatibilty.AvailableReplicas` (renamed) (deprecated) | (removed)                           |
| `FailureReason` (deprecated)         | `BackCompatibilty.FailureReason` (renamed) (deprecated)     | (removed)                           |
| `FailureMessage` (deprecated)        | `BackCompatibilty.FailureMessage` (renamed) (deprecated)    | (removed)                           |
| `Conditions`                         | `BackCompatibilty.Conditions` (renamed) (deprecated)        | (removed)                           |
| `ExperimentalConditions` (new)       | `Conditions` (renamed)                                      | `Conditions`                        |

> | `UpToDateReplicas` (new)             | `UpToDateReplicas`                                          | `UpToDateReplicas`                  |

Reviewer: Why is 'UpToDateReplicas' not prefixed with 'Experimental' yet it is also new?

Reviewer 2: The reason why we have to use experimental is that we have naming conflicts. E.g. with v1beta2 we want to have the same condition field with a different type. To be able to already introduce something now and allow a smooth migration we need a different name for v1beta1.

We only did this "name migration" for fields with conflicts. `UpToDateReplicas` is an entirely new field so there is no conflict to handle

Author: resolving this in favour of https://github.com/kubernetes-sigs/cluster-api/pull/10897/files#r1688511927 where we are discussing moving all experimental/new fields under `v1Beta2` (also trying to keep the number of open threads under control 😅 )

- This proposal is using `UpToDateReplicas` instead of `UpdatedReplicas`; This is a deliberated choice to avoid 
  confusion between update (any change) and upgrade (change of the Kubernetes versions).
- Also `AvailableReplicas` will determine Machine's availability by reading Machine.Available condition instead of
  computing availability as of today, however in this case the semantic of the field is not changed

TODO: check `FullyLabeledReplicas`, do we still need it?

#### MachineSet (New)Conditions

| Condition        | Note                                                                                                             |
|------------------|------------------------------------------------------------------------------------------------------------------|
| `ReplicaFailure` | This condition surfaces issues on creating a Machine replica in Kubernetes, if any. e.g. due to resource quotas. |
| `MachinesReady`  | This condition surfaces detail of issues on the controlled machines, if any.                                     |
| `ScalingUp`      | True if available replicas < desired replicas                                                                    |

> | `ScalingDown`    | True if replicas > desired replicas                                                                              |

Author: chatted about this, 
we are going to count both at MS level as well as upper levels (not aggregating)
that means that at any level we are either scaling up or down

| `UpToDate`       | True if all the Machines controlled by this MachineSet are up to date (replicas = upToDate replicas)             |

> | `Remediating`    | True if there is at least one Machine controlled by this MachineSet that is not passing health checks            |

Reviewer 8: "Remediating" is a really _active_ sounding condition: it communicates to me that a problem is being fixed.

But the description of it is a little different: a problem has been identified.

It seems like these should either be two distinct conditions to represent (a) that a machine is not passing health checks and (b) that remediation is in progress. Or align the name and description to either (a) or (b).

Author: In Cluster API the fact that a problem has been identified by a MachineSet also implies that remediation has been attempted.

Considering this, I think the current description is ok because it represents what is actually happening; also it relates to something the user can validate by simply looking at machines and it is similar/consistent with other conditions e.g. `ScalingDown` condition being described as true if replicas > desired replicas.

But happy to consider improvements to the descriptions, if any.

What instead I don't think we should do is to introduce another condition; this will probably introduce more confusion than benefits. 

Expanding a little bit on this last point, I don't think it matters to users if something happens one or two reconcile loops after the trigger for this action is detected, and in case the action cannot happen/is intentionally delayed, this should surface in the condition message, e.g. remediation not possible because...

Author: Now the description of the condition is "This condition surfaces details about ongoing remediation of the controlled machines, if any", which should address any doubt.

| `Deleted`        | True if MachineSet is deleted; Reason can be used to observe the cleanup progress when the resource is deleted   |
| `Paused`         | True if this MachineSet or the Cluster it belongs to are paused                                                  |

> To better evaluate proposed changes, below you can find the list of current MachineSet's conditions:
> Ready, MachinesCreated, Resized, MachinesReady.

Notes:
- MachineSet conditions are intentionally mostly consistent with MachineDeployment conditions to help users troubleshooting .
- MachineSet is considered as a sort of implementation detail of MachineDeployments, so it doesn't have its own concept of availability.
  Similarly, this proposal is dropping the notion of MachineSet readiness because it is preferred to let users focusing on Machines readiness.
- `Remediating` for older MachineSet sets will report that remediation will happen as part of the regular rollout.
- `UpToDate` condition initially will be `false` for older MachineSet, `true` for the current MachineSet; however in

>   the future the latter might evolve in case Cluster API will start supporting in-place upgrades.

Author: clarified, PTAL

    // +optional
    ReadyReplicas int32 `json:"readyReplicas"`

    // The number of available replicas for this MachineDeployment. A machine is considered available when Machine's Available condition is true.
    // +optional
    AvailableReplicas int32 `json:"availableReplicas"`

    // The number of up-to-date replicas targeted by this deployment.
    // +optional
    UpToDateReplicas int32 `json:"upToDateReplicas"`

    // Represents the observations of a MachineDeployment's current state.
    // +optional
    // +listType=map
    // +listMapKey=type
    Conditions []metav1.Condition `json:"conditions,omitempty"`

    // Other fields...
    // NOTE: `Phase`, `FailureReason`, `FailureMessage` fields won't be there anymore
}
```

| v1beta1 (current)                    | v1beta2 (tentative Q1 2025)                                 | v1beta1 removal (tentative Q1 2026) |
|--------------------------------------|-------------------------------------------------------------|-------------------------------------|
| `UpdatedReplicas`                    | `UpToDateReplicas` (renamed)                                | `UpToDateReplicas`                  |

> | `ExprimentalReadyReplicas` (new)     | `ReadyReplicas` (renamed)                                   | `ReadyReplicas`                     |

Reviewer: typo: Exprimental...

| other fields...                      | other fields...                                             | other fields...                     |

Notes:
- The `BackCompatibilty` struct is going to exist in v1beta2 types only until v1beta1 removal (9 months or 3 minor releases after v1beta2 is released/v1beta1 is deprecated, whichever is longer).
  Fields in this struct are used for supporting down conversions, thus providing users relying on v1beta1 APIs additional buffer time to pick up the new changes.

#### MachineDeployment (New)Conditions

| Condition        | Note                                                                                                                                                                                                                                                   |
|------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | `Available`      | True if the MachineDeployment has minimum availability according to parameters specified in the deployment strategy, e.g. If using RollingUpgrade strategy, availableReplicas must be greater or equal than desired replicas - MaxUnavailable replicas |  

Reviewer 8: > availableReplicas must be greater or equal than desired replicas - MaxUnavailable replicas

Is the `RollingUpgrade` strategy the right source of this information? I may only want one node to be intentionally unavailable at a time during a planned rollout, but expect to be able to tolerate two nodes being unavailable (e.g., in case some failure occurs during an upgrade). If I've planned for that, maybe I wouldn't want to see `Available: false` if I've got two nodes unavailable?

I also wonder about possible interactions here with scaling operations, including those triggered by Cluster Autoscaler: if the desired number of replicas increases a bunch, seeing `ScalingUp: true` and `Available: false` seems like it could be unnecessarily alarming. From a glance, it seems like something bad happened during the scale up, rather than an expected consequence of a scale up.

Author: Sharing a few initial thoughts.

- MD RollingUpgrade strategy IMO is the current source of this information, because this is the single point of control for this group of machines and everything below should be considered ephemeral
- The idea of unbalanced/more sophisticated RollingUpgrade strategies could be something interesting to think about, even if the downside of it is that it adds complexity to a behaviour that is nicely consistent with K8s Deployments.
- `ScalingUp: true` and `Available: false` does not seem alarming to me _if_ this is considered together with current and desired replicas, because the last two info should give you an hint about the fact that someone or something asked for additional capacity (and those info are all available in the same place)

However, frankly speaking, even if interesting, discussing how MD performs rolling upgrades + if and how we should improve this seems a little bit out of scope of this proposal (also MD.status.Conditions[Available] one of the conditions we are not changing in this proposal 😅 ).

But If we want to continue this discussion happy to do it on a separated issue

Reviewer 2: After thinking a bit about this.

I could also see the case for deciding available just based on if "available replicas" >= "desired replicas".

I think this is already the case if someone uses MaxUnavailable = 0, or if folks would use a different strategy.

Also really not sure if the updateStrategy should be the source of truth for the decision what "available" means on a MachineDeployment, it feels a bit out of scope for the updateStrategy.

Reviewer 2: But agree, the current definition in this proposal is our current implementation. Maybe worth a follow-up on the umbrella issue though

Author: Resolving, eventually we can follow up on ideas for improvements on rolling upgrades on a separated issue

> | `ReplicaFailure` | This condition surfaces issues on creating a MachineSet replica in Kubernetes, if any. e.g. due to resource quotas.                                                                                                                                    |

Author: Yes, added a note explaining this

| `MachinesReady`  | This condition surfaces detail of issues on the controlled machines, if any.                                                                                                                                                                           |
| `ScalingUp`      | True if available replicas < desired replicas                                                                                                                                                                                                          |

> | `ScalingDown`    | True if replicas > desired replicas                                                                                                                                                                                                                    |

Reviewer 8: If going through a rollout with something like `MaxSurge = 3`, is `ScalingDown` going to report `True`? (And vice versa with `MaxUnavailable` and `ScalingUp`?)

Author: If going through a rollout with something like MaxSurge = 3 and MaxUnaivailable=0, ScalingDown will not show up (it depends by both settings, not only by one).
Same for ScalingUp

Author: Hopefully now it is clear, resolving

Below the Go types, you can find a summary table that also shows how changes will be rolled out according to K8s deprecation rules.

```golang
type ClusterStatus struct {

    // Initialization provides observations of the Cluster initialization process.
    // NOTE: fields in this struct are part of the Cluster API contract and are used to orchestrate initial Cluster provisioning.
    // The value of those fields is never updated after provisioning is completed.
    // Use conditions to monitor the operational state of the Cluster's BootstrapSecret.
    // +optional
    Initialization *MachineInitializationStatus `json:"initialization,omitempty"`
    
    // Represents the observations of a Cluster's current state.
    // +optional
    // +listType=map
    // +listMapKey=type
    Conditions []metav1.Condition `json:"conditions,omitempty"`
    
    // ControlPlane groups all the observations about Cluster's ControlPlane current state.
    // +optional
    ControlPlane *ClusterControlPlaneStatus `json:"controlPlane,omitempty"`
    
    // Workers groups all the observations about Cluster's Workers current state.
    // +optional

>     Workers *ClusterControlPlaneStatus `json:"workers,omitempty"`

Reviewer 2: Also it shouldn't be "WorkersPlane" (plane doesn't make sense in that struct name :))

notes:
- The `BackCompatibilty` struct is going to exist in v1beta2 types only until v1beta1 removal (9 months or 3 minor releases after v1beta2 is released/v1beta1 is deprecated, whichever is longer).
  Fields in this struct are used for supporting down conversions, thus providing users relying on v1beta1 APIs additional buffer time to pick up the new changes.

##### Cluster (New)Conditions

| Condition                 | Note                                                                                                                                                                                                                                                                                                                  |
|---------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | `Available`               | True if Cluster `RemoteConnectionProbe` is true, if Cluster's control plane `Available` condition is true, if all MachineDeployment and MachinePool's `Available` condition are true; if conditions are defined in `spec.availabilityGates`, those conditions should be true as well for the Cluster to be available. |

Reviewer 8: Consider a Cluster with 3 MachineDeployments, each in a separate failure domains, intended to tolerate the loss of any one of those failure domains.

> all MachineDeployment and MachinePool's `Available` condition are true

I think it'd be confusing for that Cluster to report `Available: false` when one failure domain goes down. It seems helpful to be able to distinguish between "a failure domain is down and my cluster is still available as I expected" (`Available: true`?) and "outage of one failure domain has caused some cascading issue causing the cluster to fail" (`Available: false`?).

Reviewer 2: Not sure how we would model this :)

Author: Agree with @sbueringer 
Also, adding a little bit on top, we have several issues in our backlog about improving failure domains, and I think we should look at this area in a more holistic way addressing not only availability, but also rebalancing, removal/addition of failure domains, maintenance more, probably more (which is something out of the scope of this proposal).

Author: Resolving, eventually we can follow up on failure domain improvements on a separated issue (or on one of the existing issues about the same topic)

> | `ControlPlaneInitialized` | True when the Cluster's control plane is functional enough to accept requests. This information is usually used as a signal for starting all the provisioning operations that depends on a functional API server, but do not require a full HA control plane to exists.                                               |

Reviewer 8: I'm not clear on the relationship between this and `ClusterInitializationStatus.ControlPlaneInitialized`. When should I look at one vs. the other?

When would I look at this instead of `ControlPlaneAvailable`?

Author: status.Initialized.ControlPlaneInitialized and status.conditions[ControlPlaneInitialized] are the same signal, but conditions offer an overall status on the control plane which instead do not exist under status.Initialized (you can look at both, but IMO using conditions should be preferred)

More context:

status.Initialized.ControlPlaneInitialized  comes from the ClusterAPI contract, and we are preserving this field because proposal is not going to change the fact that Cluster API contract uses flags to orchestrate initial provisioning (see latest paragraphs I have added to the doc for more context, also divide and conquer 😉)

status.conditions[ControlPlaneInitialized] has been added because it can be used in combination with `kubectl wait` when writing simple scripts to automate processes around cluster provisioning

Author: Resolving, hopefully it is clear now

> | `RemoteConnectionProbe`   | True when control plane can be reached; in case of connection problems, the condition turns to false only if the the cluster cannot be reached for 40s after the first connection problem is detected (or whatever period is defined in the `--remote-connection-grace-period` flag) the cluster cannot be reached    |

Reviewer 2: Sounds like a good idea. Default (via flag) + overwrite possible via Cluster CR?

> | `RemoteConnectionProbe`   | True when control plane can be reached; in case of connection problems, the condition turns to false only if the the cluster cannot be reached for 40s after the first connection problem is detected (or whatever period is defined in the `--remote-connection-grace-period` flag) the cluster cannot be reached    |

Author: No strong objections, but also in this case I would prefer to implement the feature first, and then implement support for exceptions at cluster level in a follow-up.

Let's hear other opinions
Please add 👍 if you are in favor --remote-connection-grace-period per cluster ASAP 👎 if you are in favor of having this as a follow-up

> | `RemoteConnectionProbe`   | True when control plane can be reached; in case of connection problems, the condition turns to false only if the the cluster cannot be reached for 40s after the first connection problem is detected (or whatever period is defined in the `--remote-connection-grace-period` flag) the cluster cannot be reached    |

Reviewer 2: I think follow-up would be better. Thinking about this a bit more. In general I think if we're taking about latency (how long our calls to the workload cluster takes) I could imagine that there are differences, but I would assume there is some sort of threshold until we can deal with request latency (we can compare with the timeouts we use today when communicating with the workload cluster)

But I think this configuration here is not about latency (i.e. not a request timeout), it's about how long we weren't able to communicate with the wl cluster (after trying it multiple times). I would imagine that even with high latency we should be able to communicate with wl clusters after >= 10 seconds if they are up.

> | `RemoteConnectionProbe`   | True when control plane can be reached; in case of connection problems, the condition turns to false only if the the cluster cannot be reached for 40s after the first connection problem is detected (or whatever period is defined in the `--remote-connection-grace-period` flag) the cluster cannot be reached    |

Author: Tracked in https://github.com/kubernetes-sigs/cluster-api/issues/11105

| `InfrastructureReady`     | Mirror of Cluster's infrastructure `Ready` condition                                                                                                                                                                                                                                                                  |
| `ControlPlaneAvailable`   | Mirror of Cluster's control plane `Available` condition                                                                                                                                                                                                                                                               |

> | `WorkersAvaiable`         | Summary of MachineDeployment and MachinePool's `Available` condition                                                                                                                                                                                                                                                  |

Reviewer: typo: `Avaiable`

| `TopologyReconciled`      |                                                                                                                                                                                                                                                                                                                       |
| `ScalingUp`               | True if available replicas < desired replicas                                                                                                                                                                                                                                                                         |
| `ScalingDown`             | True if replicas > desired replicas                                                                                                                                                                                                                                                                                   |

> | `UpToDate`                | True if all the Machines controlled by this Cluster are up to date (replicas = upToDate replicas)                                                                                                                                                                                                                     |

Author: It is correct (up to date at cluster level signals all the machines are in final state)
